Use feature flags for extern definitions

marzipankaiser opened the pull request:
This implements feature flags for externs.

# Syntax
```effekt
extern def infixAdd(a: Int, b: Int): Int =
     js "${a} + ${b}"
     llvm { someEffektStmt }
     default "${a} + ${b}"
```
(Note: You might not want to use `default` like this... Just an example.)

The old syntax should still work and is equivalent to having just a `default` clause.

This also exists for includes like this:
```effekt
extern js """
...
"""
extern include js " ... "
```

# Notes about implemented behaviour
Current behaviour is:
- Extern includes with non-matching flags are ignored
- Extern defs without matching flag (for current backend) give out a warning during naming (for now; ideally: Check if it is actually used!)
- Extern defs with an unannotated/`default` **string** will give out a warning

# Future work
- Output warnings only if we (transitively) use unsupported externs from main
- Move generation of warnings from `Namer` into a separate phase, potentially with some other smaller checks. Alternatively: Move to `ResolveExternDefs`.
- Allow using interfaces/functions passed as block parameters (currently fails with a type error due to captures).

jiribenes commented on effekt/shared/src/main/scala/effekt/FeatureFlag.scala:
```diff
@@ -1,4 +1,4 @@
-package effekt.symbols
+package effekt
```
_Why does this help?_ 🤔

This is very strange.

marzipankaiser commented:
Design decision reg. Effekt code in extern definitions:
Currently, the syntax is just `(expr)` (forced parentheses for parsing) instead of a string.
They (all of them) are passed through to the backends and transformed on the way, the backend can then choose.

Upside: The backends can do whatever they want with it.

Downside: This has to be handled in the backends, which is non-trivial at least in LLVM.

Alternative: Choose the "correct" extern definition in `core`. Then make the pseudo-extern definitions normal definitions.
(Pro/con as above but reversed)

b-studios commented:
I don't think I understand your proposal. Do you have an example of how this looks like?

marzipankaiser commented:
Note to self: Add a simple phase on core to select & desugar extern defs (the backends can decide to use it or not)

marzipankaiser commented:
Changed:
- Now uses `{ stmt }` instead of `( expr )` for effekt-defined externs; this makes the second point easier and seems more natural to me
- Added a Core -> Core phase that selects the correct definitions and makes them "normal" definitions if they are defined using Effekt code

marzipankaiser commented on effekt/shared/src/main/scala/effekt/Namer.scala:
```diff
@@ -288,13 +290,20 @@ object Namer extends Phase[Parsed, NameResolved] {
       }
 
-    case f @ source.ExternDef(capture, id, tparams, vparams, bparams, ret, body) =>
+    case f @ source.ExternDef(capture, id, tparams, vparams, bparams, ret, bodies) =>
+      if (!bodies.supportedByFeatureFlags(Context.compiler.supportedFeatureFlags)) {
+        Context.warning(s"Extern definition ${id} is not supported by this backend.")
+      }
```
This is where the warnings are currently generated.

jiribenes replied:
Looks great! 👍
For clarity, would it be possible to also dump why exactly it's not supported? (something like `Extern definition ${id} is not supported as it requires feature flags ${featureFlags}, but the current backend only supports ${backendFeatureFlags}`)

marzipankaiser commented on effekt/shared/src/main/scala/effekt/Typer.scala:
```diff
@@ -829,11 +830,98 @@ object Typer extends Phase[NameResolved, Typechecked] {
+              sym.bparams foreach Context.bind
+
+              // to subtract the capabilities, which are only inferred bottom up, we need a **second** unification variable
+              val inferredCapture = Context.freshCaptVar(CaptUnificationVar.AnonymousFunctionRegion(BlockLiteral(tps, vps, bps, body)))
```
Apart from the role here, all of this code is just copy-pasted from the case for `FunDef`.

b-studios replied:
That's a lot of code. Can we somehow share?

marzipankaiser replied:
We probably could. Thinking about it, frankly, I'm not sure this is the correct way to check those.
We probably want at least "internal" type checking, but I'm not quite certain how much we want at the boundary.

Will look at it again once I find the time.

marzipankaiser commented on effekt/shared/src/main/scala/effekt/core/ResolveExternDefs.scala:
```diff
@@ -0,0 +1,50 @@
+import effekt.context.Context
+import effekt.source.FeatureFlag
+
+/**
+ * [[Phase]] that resolves feature-flag annotated externs to unique extern (string-based) defs or new definitions.
+ *
+ * @param supportedFeatureFlags are a *prioritized* list (i.e., earlier ones always match).
+ * Unsupported externs are replaced with a definition according to [[defaultExternBody]].
+ */
+case class ResolveExternDefs(supportedFeatureFlags: List[String]) extends Phase[CoreTransformed, CoreTransformed] {
```
This is the phase resolving extern defs

b-studios commented on effekt/shared/src/main/scala/effekt/Parser.scala:
```diff
@@ -163,20 +173,27 @@ class EffektParsers(positions: Positions) extends EffektLexers(positions) {
+            )) ^^ {
+                case ff ~ body => ExternBody.StringExternBody(ff, body)
+            }
+    | featureFlag ~ ("{" ~> stmt <~ "}") ^^ { case ff ~ body => ExternBody.EffektExternBody(ff, body) }
```
```suggestion
    | featureFlag ~ ("{" ~> stmts <~ "}") ^^ { case ff ~ body => ExternBody.EffektExternBody(ff, body) }
```

marzipankaiser commented:
At some point we might want to move the warnings at
https://github.com/effekt-lang/effekt/blob/5646f6731f7983dcead702a6e60321f9cc9186c2/effekt/shared/src/main/scala/effekt/Namer.scala#L294-L304
and
https://github.com/effekt-lang/effekt/blob/5646f6731f7983dcead702a6e60321f9cc9186c2/effekt/shared/src/main/scala/effekt/Namer.scala#L385-L389
from `Namer` to a separate phase with more of those smaller checks.
Maybe there are also other checks that would fit there?

marzipankaiser commented:
Tests are failing because of tests that use definitions like
https://github.com/effekt-lang/effekt/blob/5646f6731f7983dcead702a6e60321f9cc9186c2/examples/pos/capture/resources.effekt#L3
which will potentially never work on some backends. 
Potentially some can be rewritten to use Effekt definitions instead (i.e., `extern def ... = else { ... }`).

marzipankaiser commented on effekt/shared/src/main/scala/effekt/Compiler.scala:
```diff
@@ -66,6 +66,17 @@ export PhaseResult.*
+/**
+ * Constants for common feature flags (not specific to one generator)
+ */
+object FeatureFlag {
+  /** For backends that translate to a programming language supporting at least:
+   * - directly splicing in variables ("${...}" works)
+   * - "simple" literals like "42"
+   */
+  val simpleAtom: String = "simpleAtom"
```
Here, I now define common feature flags (currently one).
`simpleAtom` is meant to be supported by backends that translate such that both simple literals (ints like "42") and variable names (mostly for "${x}") are allowed.

marzipankaiser replied:
(Removed)

b-studios commented on examples/pos/capture/ffi_blocks.effekt:
```diff
@@ -1,6 +1,6 @@
-extern io def something {f: Foo}: Int = "${box f}"
-extern io def createFoo(): Foo at {global} = "42"
+extern io def something {f: Foo}: Int = simpleAtom "${box f}"
+extern io def createFoo(): Foo at {global} = simpleAtom "42"
```
This is just a bad test (already has been), not your fault

marzipankaiser replied:
I tried to fix it up, and it found a bug: We cannot actually call any operations on `f` since the capture is not allowed by our current implementation of Typer here.
Just allowing this in typer, however, leads to crashes in the backends (more precisely: undefined variables). I think this might be related to not passing the captures to the extern 🤔.

b-studios commented on examples/pos/capture/mbed.effekt:
```diff
@@ -9,10 +9,10 @@ extern resource deep: Builtin
-extern {out}   def useOut(): Unit = "0"
-extern {block} def performBlocking(): Unit = "0"
-extern {deep}  def blockDeepSleep(): Unit = "0"
+extern {in}    def useIn(): Unit = simpleAtom "0"
```
```suggestion
extern {in}    def useIn(): Unit = else { 0 }
```

marzipankaiser commented:
`else` -> `default`

b-studios commented on effekt/shared/src/main/scala/effekt/Typer.scala:
```diff
@@ -829,11 +830,22 @@ object Typer extends Phase[NameResolved, Typechecked] {
         d.symbol.bparams foreach Context.bind
 
-        body.args.foreach { arg => checkExpr(arg, None) }
+        flowingInto(Context.lookupCapture(id.symbol.asBlockSymbol)) {
```
would the following work?
```suggestion
        flowingInto(Context.lookupCapture(d.symbol)) {
```

marzipankaiser replied:
Should work, will change.

b-studios commented on effekt/shared/src/main/scala/effekt/core/PrettyPrinter.scala:
```diff
@@ -59,9 +60,17 @@ object PrettyPrinter extends ParenPrettyPrinter {
+
+  def toDoc(ff: FeatureFlag): Doc = ff match {
+    case FeatureFlag.NamedFeatureFlag(name) => name
+    case FeatureFlag.Default => "else"
```
Nice, here it is already called `Default` :)

marzipankaiser replied:
Just noticed so too when renaming 😅

b-studios commented on effekt/shared/src/main/scala/effekt/core/Tree.scala:
```diff
@@ -125,8 +126,22 @@ case class Property(id: Id, tpe: BlockType) extends Tree
+}
+object ExternBody {
+  case class StringExternBody(featureFlag: FeatureFlag, contents: Template[Pure]) extends ExternBody
+  case class EffektExternBody(featureFlag: FeatureFlag, body: Stmt) extends ExternBody
```
Potentially risking conflicts with `String` this could be shortened to
```scala
enum ExternBody extends Tree {
  def featureFlag: FeatureFlag
  
  case String(featureFlag: FeatureFlag, contents: Template[Pure])
  case EffektExternBody(featureFlag: FeatureFlag, body: Stmt)
}
```

marzipankaiser replied:
I'm not sure it's worth it given the potential conflicts. In particular, adding any helper methods on `ExternBody` that use `String`s as parameters/return values would become awkward (and might not be that unlikely).
Maybe we can also come up with a better name (I couldn't, hence the not entirely helpful one... -s, actually).

marzipankaiser commented on examples/pos/liftinference.effekt:
```diff
@@ -5,7 +5,7 @@ interface Foo {
 }
 
-extern control def fancy { f: Unit => Int }: Int = "foo"
+extern control def fancy { f: Unit => Int }: Int = simpleAtom "foo"
```
Here, using `default { ... }` instead would potentially make the test less useful - after all, there probably was a reason for using an extern here, and this will never reach `LiftInference` then (will be desugared in core already).

b-studios replied:
Oh that is true. But we could also run it only in one of the lift backends.

marzipankaiser replied:
I'll probably just duplicate it. 🤔

marzipankaiser commented:
Maybe it would be better to move the Resolving further to the front after all.
I would still prefer it to be after Typing, but we could do it in `Middleend` on `Typechecked`.

marzipankaiser commented:
The approach of adding it after Typer does not work properly either: The calling-convention issue hinges on the symbol type used in `core.Transformer`:
https://github.com/effekt-lang/effekt/blob/ee905df3182f203730cec00d6cffd3472b84cbdb/effekt/shared/src/main/scala/effekt/core/Transformer.scala#L635-L638

b-studios commented on effekt/shared/src/main/scala/effekt/core/Parser.scala:
```diff
@@ -69,13 +69,11 @@ class CoreParsers(positions: Positions, names: Names) extends EffektLexers(posit
+        Extern.Def(id, tparams, cparams, vparams, bparams, result, captures, body match {
           case ff ~ (body: String) =>
-            ExternBody.StringExternBody(ff, Template(List(body), Nil))
+            ExternBody(ff, Template(List(body), Nil))
```
Beautiful! Thanks, that's much better

b-studios commented on effekt/shared/src/main/scala/effekt/core/Transformer.scala:
```diff
@@ -117,18 +117,19 @@ object Transformer extends Phase[Typechecked, CoreTransformed] {
     case f @ source.ExternDef(pure, id, _, vps, bps, _, bodies) =>
       val sym@ExternFunction(name, tps, _, _, ret, effects, capt, _) = f.symbol
       assert(effects.isEmpty)
+      assert(bodies.length == 1, "Externs are resolved before core")
```
The `.head` below is a bit hidden. Maybe refactor to

```
val tBody = bodies match {
  case source.ExternBody.StringExternBody(ff, body) :: Nil => ...
  case _ => Context.panic("Externs should be resolved and desugared before core.Transformer")
}
```
?

marzipankaiser replied:
Good idea.

b-studios commented on effekt/shared/src/main/scala/effekt/core/Tree.scala:
```diff
@@ -126,16 +126,11 @@ case class Property(id: Id, tpe: BlockType) extends Tree
  */
 enum Extern extends Tree {
-  case Def(id: Id, tparams: List[Id], cparams: List[Id], vparams: List[Param.ValueParam], bparams: List[Param.BlockParam], ret: ValueType, annotatedCapture: Captures, bodies: List[ExternBody])
+  case Def(id: Id, tparams: List[Id], cparams: List[Id], vparams: List[Param.ValueParam], bparams: List[Param.BlockParam], ret: ValueType, annotatedCapture: Captures, body: ExternBody)
```
nice!

b-studios commented on effekt/shared/src/main/scala/effekt/generator/js/TransformerMonadic.scala:
```diff
@@ -183,22 +179,14 @@ trait TransformerMonadic extends Transformer {
-          inlineExtern(vargs, extern.vparams, body)
-        case ExternBody.EffektExternBody(_, body) => sys error "Effekt extern body should have been removed"
-      }
+      inlineExtern(vargs, extern.vparams, extern.body.contents)
```
Much, much better!

b-studios commented on effekt/shared/src/main/scala/effekt/Compiler.scala:
```diff
@@ -66,6 +66,12 @@ export PhaseResult.*
+ * Constants for common feature flags (not specific to one generator)
+ */
+object FeatureFlag {
+}
```
Do we still need this now it is empty?

marzipankaiser replied:
Not really, removing it.

b-studios commented on effekt/shared/src/main/scala/effekt/Namer.scala:
```diff
@@ -288,13 +290,28 @@ object Namer extends Phase[Parsed, NameResolved] {
+    case f @ source.ExternDef(capture, id, tparams, vparams, bparams, ret, bodies) =>
+      if (!bodies.supportedByFeatureFlags(Context.compiler.supportedFeatureFlags)) {
+        val featureFlags = bodies.map(_.featureFlag)
+        Context.warning(pp"Extern definition ${id} is not supported as it is only is defined for feature flags ${featureFlags.mkString(", ")}," +
```
```suggestion
        Context.warning(pp"Extern definition ${id} is not supported as it is only defined for feature flags ${featureFlags.mkString(", ")}," +
```

b-studios commented on effekt/shared/src/main/scala/effekt/generator/js/TransformerMonadic.scala:
```diff
@@ -35,7 +37,8 @@ object TransformerMonadicWhole extends TransformerMonadic {
       !hasBlockParameters && !isControlEffecting
-    case Extern.Include(contents) => false
+    case Extern.Include(ff, contents) if ff.matches(jsFeatureFlags) => false
+    case Extern.Include(ff, contents) => true // it's nothing anyway
```
This I don't understand. Before we didn't inline those at all.

b-studios commented on effekt/shared/src/main/scala/effekt/generator/js/TransformerMonadic.scala:
```diff
@@ -110,11 +113,13 @@ trait TransformerMonadic extends Transformer {
+    case Extern.Include(ff, contents) if ff.matches(jsFeatureFlags) =>
       js.RawStmt(contents)
+
+    case Extern.Include(_, _) => js.RawStmt("") // ignore, not meant for us
```
Can this still happen? I thought selection happens earlier and only leaves those that we care about.

marzipankaiser replied:
Or just drop the check altogether. I'll do the latter.

b-studios commented on effekt/shared/src/main/scala/effekt/generator/llvm/Transformer.scala:
```diff
@@ -40,12 +42,14 @@ object Transformer {
+      case machine.Include(ff, content) if ff.matches(llvmFeatureFlags) =>
         Verbatim(content)
+      case machine.Include(ff, content) =>
+        Verbatim("") // ignore, not meant for us
```
Same as above, is this necessary?

b-studios commented:
I reviewed the Scala source changes and they look mostly good to me. I only have a few remarks:

1. https://github.com/effekt-lang/effekt/pull/427#discussion_r1572385267
2. https://github.com/effekt-lang/effekt/pull/427#discussion_r1572605230 (and for other backends)
3. https://github.com/effekt-lang/effekt/pull/427#discussion_r1572603819

marzipankaiser commented:
(interpreted that as approval)
